### DISCUSSION TITLE
Testing React Components with Xstate useMachine Hook & services

### DISCUSSION THREAD

**Discussion by dsathyakumar** [statelyai/xstate#2303] [Q&A] [answered]
Hello! 👋 

I'd like to know if there are any sample docs around how to test React Components with Xstate (with React Testing library)  `useMachine` hook & cases like invoking services (services that make `fetch` / `XmlHttpRequest` calls to GET / POST stuff).
I'd like to know how this can be stubbed / mocked and tested (with spies) etc.,. 

CC - @davidkpiano @Andarist

**Comment by davidkpiano** [comment, marked as answer]
Check out how the tests are done here: https://github.com/davidkpiano/xstate/blob/main/packages/xstate-react/test/useMachine.test.tsx

**Reply by dsathyakumar** [reply]
@davidkpiano The example you shared doesn't indicate how a service call can be mocked. Do you have any examples around it ?
For example, a service call to upload a file etc., I could not find much examples about how the call can be mocked.

**Reply by dsathyakumar** [reply]
CC - @Andarist

**Reply by davidkpiano** [reply]
> @davidkpiano The example you shared doesn't indicate how a service call can be mocked. Do you have any examples around it ?
> For example, a service call to upload a file etc., I could not find much examples about how the call can be mocked.

If you have access to the machine, you can mock it. Otherwise, it's a closure, and it's hard to mock closed-in values in general.

For example, your component might accept a machine as props, and you can override the services via `machine.withConfig({ services: ... })` which makes for straightforward mocking.

If this isn't possible, I recommend using a library like [Mock Service Worker](https://mswjs.io/) to get the job done.

**Reply by dsathyakumar** [reply]
Thanks @davidkpiano I did try the withConfig method, but it seems buggy. I wasn't able to override my original implementation.
I created a repl here:
[xstate-with-config-issue](https://replit.com/@dsathyakumar/ScornfulTemporalCallback )
It still continues to use the old `actions` only and updates the context value per the old action.

**Reply by dsathyakumar** [reply]
Im on xstate@^4.17.1

**Reply by dsathyakumar** [reply]
nvm i got it .. it had to be reassigned!. Thanks @davidkpiano 👍
